File: 4/Java/Solution.java

public class Solution {
    public int findMaxSumOfrray(int[] array) {
        if (array == null) {
            return 0;
        }
        int sum = array[0];
        int tempsum = array[0];
        for (int i = 1; i < array.length; i++) {
            tempsum = Math.max(tempsum + array[i], array[i]);
            sum = Math.max(tempsum, res);
        }
        return sum;
    }

    public static void main(String[] args) {
        int[] array = {1,-2,3,10,-4,7,2,-5};
        System.out.println(new Solution().findMaxSumOfrray(array));
    }
}

File: 4/Python/Solution.py
# -*- coding:utf-8 -*-

class Solution:
    #动态规划方法，每次计算一小块最优解，计算下一个时，把之前的解当做一个整体再进行计算
    def FindGreatestSumOfSubArray(self, array):
        res = max(array)
        temp = array[0]
        for i in array:
            temp = max(temp + i, i)
            res = max(res, temp)
            
        return res

print(Solution().FindGreatestSumOfSubArray([1, -2, 3, 10, -4, 7, 2, -5]))

